Fill in sidebarAction / sidePanel WebProcess <=> UIProcess messaging stubs.

AriYoung00 opened the pull request:
#### b2e851ada7eee7b13ffde9f5d2d6189461732869
<pre>
Fill in sidebarAction / sidePanel WebProcess &lt;=&gt; UIProcess messaging stubs.
<a href="https://webkit.org/b/277688">https://webkit.org/b/277688</a>
<a href="https://rdar.apple.com/133073249">rdar://133073249</a>

Reviewed by Timothy Hatcher.

This PR fills in most of the message senders and receivers which pertain
to sidePanel / sidebarAction. Additionally, it adds several helper
utilities to aid in parameter parsing for sidePanel / sidebarAction.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::toExpected): Utility function to convert an optional&lt;T&gt;
to an Expected&lt;T, WebExtensionError&gt;
(WebKit::getSidebarWithIdentifiers): Utility function to validate window
/ tab identifiers then return the sidebar object associated with said
identifiers
(WebKit::getOrCreateSidebarWithIdentifiers): Utility function to
validate window / tab identifiers and create a sidebar object
associated with said identifiers if necessary
(WebKit::WebExtensionContext::openSidebarForTab): Utility function to
open a sidebar object
(WebKit::WebExtensionContext::closeSidebarForTab): Utility function to
close a sidebar object
(WebKit::WebExtensionContext::sidebarOpen): Fill in message receiver for
sidebarOpen
(WebKit::WebExtensionContext::sidebarClose): Fill in message receiver for
sidebarClose
(WebKit::WebExtensionContext::sidebarIsOpen): Fill in message receiver for
sidebarIsOpen
(WebKit::WebExtensionContext::sidebarToggle): Fill in message receiver for
sidebarToggle
(WebKit::WebExtensionContext::sidebarSetIcon): Update stub with FIXME
(WebKit::WebExtensionContext::sidebarGetTitle): Fill in message receiver for
sidebarGetTitle
(WebKit::WebExtensionContext::sidebarSetTitle): Fill in message receiver for
sidebarSetTitle
(WebKit::WebExtensionContext::sidebarGetOptions): Fill in message receiver for
sidebarGetOptions
(WebKit::WebExtensionContext::sidebarSetOptions): Fill in message receiver for
sidebarSetOptions
(WebKit::WebExtensionContext::isSidebarMessageAllowed): Gate sidebar
messages behind sidebar feature flag
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
(WebKit::parseTabIdentifier): Utility function to parse and validate the
tab identifier received through a sidePanel details object
(WebKit::parseWindowIdentifier): Utility function to parse and validate
the window identifier received through a sidePanel details object
(WebKit::serializeSidebarParameters): Utility function to serialize a
WebExtensionSidebarParameters struct to an NSDictionary
(WebKit::deserializeSidebarParameters): Utility function to deserialize
a WebExtensionSidebarParameters struct from an NSDictionary
(WebKit::WebExtensionAPISidePanel::getOptions): Fill in message sender
for getOptions
(WebKit::WebExtensionAPISidePanel::setOptions): Fill in message sender
for setOptions
(WebKit::WebExtensionAPISidePanel::getPanelBehavior): Update stub with
FIXME
(WebKit::WebExtensionAPISidePanel::setPanelBehavior): Update stub with
FIXME
(WebKit::WebExtensionAPISidePanel::open): Fill in message sender for
open
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
(WebKit::parseSidebarActionDetails): Utility function to parse the
`details` object passed to many sidebarAction method calls
(WebKit::parseDetailsStringFromKey): Utility function to validate the
string value of some key in a dictionary, optionally allowing it to be
NSNull
(WebKit::getIdentifiers): Utility function to retrieve whatever window
or tab identifiers are set in a variant, returning them as
std::optionals
(WebKit::WebExtensionAPISidebarAction::open): Fill in message sender for
open
(WebKit::WebExtensionAPISidebarAction::close): Fill in message sender
for close
(WebKit::WebExtensionAPISidebarAction::toggle): Fill in message sender
for toggle
(WebKit::WebExtensionAPISidebarAction::isOpen): Fill in message sender
for isOpen
(WebKit::WebExtensionAPISidebarAction::getPanel): Fill in message sender
for getPanel
(WebKit::WebExtensionAPISidebarAction::setPanel): Fill in message sender
for setPanel
(WebKit::WebExtensionAPISidebarAction::getTitle): Fill in message sender
for getTitle
(WebKit::WebExtensionAPISidebarAction::setTitle): Fill in message sender
for setTitle
(WebKit::WebExtensionAPISidebarAction::setIcon): Fill in message sender
for setIcon
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h:
(WebKit::toOptional): Utility function to attempt to access
one particular field from a variant and return it in an std::optional,
returning std::nullopt if the variant does not contain the requested
field
(WebKit::indicatesError): Utility function to determine whether a
variant indicates that an error occurred, and return the error if it
does
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::isFeatureEnabled const): Helper
function to check if a particular feature is enabled using its name.

Canonical link: <a href="https://commits.webkit.org/282026@main">https://commits.webkit.org/282026@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d9d248d82376ef58fb6dcb3b45e98a0010ebc89

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61805 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41159 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14397 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/65784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/12350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48845 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12621 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/65784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/12350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64874 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38224 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53530 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/65784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34871 "Passed tests") | | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/11281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56680 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11064 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/67513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5748 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/67513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5773 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53478 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/67513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13747 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4716 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/38043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/39139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/37788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
 namespace WebKit {
 
+template<typename T>
+static Expected<T, WebExtensionError> optionalToExpected(std::optional<T>&& opt, NSString * const errorMessage = @"value not found")
```
```suggestion
static Expected<T, WebExtensionError> toWebExtensionError(std::optional<T>&& opt, NSString * const errorMessage = @"value not found")
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+    auto maybeSidebar = getOrCreateSidebar(tab);
+    if (!maybeSidebar)
+        return;
+    auto& sidebar = maybeSidebar.value().get();
```
```suggestion
        return;

    auto& sidebar = maybeSidebar.value().get();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+    auto maybeSidebar = getOrCreateSidebar(tab);
+    if (!maybeSidebar)
+        return;
+    auto& sidebar = maybeSidebar.value().get();
```
```suggestion
        return;

    auto& sidebar = maybeSidebar.value().get();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
 void WebExtensionContext::sidebarOpen(const std::optional<WebExtensionWindowIdentifier> windowIdentifier, const std::optional<WebExtensionTabIdentifier> tabIdentifier, CompletionHandler<void(Expected<void, WebExtensionError>&&)>&& completionHandler)
 {
+    // the error below is a placeholder to fill sidebar, since we cannot instantiate it empty
+    Expected<Ref<WebExtensionSidebar>, WebExtensionError> sidebar = toWebExtensionError(nil, nil, @"Placeholder error");
```
```suggestion
    auto sidebar = toWebExtensionError(nil, nil, @"Placeholder error");
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+
 void WebExtensionContext::sidebarOpen(const std::optional<WebExtensionWindowIdentifier> windowIdentifier, const std::optional<WebExtensionTabIdentifier> tabIdentifier, CompletionHandler<void(Expected<void, WebExtensionError>&&)>&& completionHandler)
 {
+    // the error below is a placeholder to fill sidebar, since we cannot instantiate it empty
```
```suggestion
    // FIXME: the error below is a placeholder to fill sidebar, since we cannot instantiate it empty
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+                completionHandler(toWebExtensionError(nil, nil, @"could not create sidebar"));
+                return;
+            }
+            sidebar = WTFMove(maybeSidebar.value());
```
```suggestion
            }

            sidebar = WTFMove(maybeSidebar.value());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+            completionHandler(toWebExtensionError(nil, nil, @"could not create sidebar"));
+            return;
+        }
+        sidebar = WTFMove(maybeSidebar.value());
```
```suggestion
        }

        sidebar = WTFMove(maybeSidebar.value());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+            tab = window->get().activeTab();
+        else if (auto maybeTab = sidebar.value()->tab())
+            tab = RefPtr { maybeTab.value().ptr() };
+        // fall back to frontmost window
```
Not really needed.
```suggestion
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+
+    if (sidebar.value()->opensSidebar())
+        openSidebarForTab(*tab, UserTriggered::No);
+    completionHandler({ });
```
```suggestion
        openSidebarForTab(*tab, UserTriggered::No);

    completionHandler({ });
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+    }
+
+    closeSidebarForTab(*tab, UserTriggered::No);
+    completionHandler({ });
```
```suggestion
    closeSidebarForTab(*tab, UserTriggered::No);

    completionHandler({ });
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+    bool isOpen = false;
+    if (auto currentTab = window->activeTab()) {
+        if (auto currentTabSidebar = getSidebar(*currentTab))
+            isOpen = isOpen || currentTabSidebar.value()->isOpen();
```
```suggestion
            isOpen |= currentTabSidebar.value()->isOpen();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+            completionHandler(toWebExtensionError(nil, nil, @"it is not implemented"));
+            return;
+        }
+        closeSidebarForTab(*tab, UserTriggered::No);
```
```suggestion
        }

        closeSidebarForTab(*tab, UserTriggered::No);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+            completionHandler(toWebExtensionError(nil, nil, @"it is not implemented"));
+            return;
+        }
+        openSidebarForTab(*tab, UserTriggered::No);
```
```suggestion
        }

        openSidebarForTab(*tab, UserTriggered::No);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+        completionHandler(makeUnexpected(sidebar.error()));
+        return;
+    }
+    completionHandler({ });
```
```suggestion
    }

    completionHandler({ });
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+    if (auto *controller = extensionController()) {
+        WKPreferences *preferences = controller->configuration().webViewConfiguration().preferences;
+
+        for (_WKFeature *feature in [WKPreferences _features]) {
```
```suggestion
        for (_WKFeature *feature in WKPreferences._features) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,342 @@
+            if ([feature.key isEqualToString:@"WebExtensionSidebarEnabled"])
+                return [preferences _isEnabledForFeature:feature];
+        }
+    }
```
This would be good as a helper method on `WebExtensionController` that takes a string. We will likely want this for other features in the future.

```cpp
bool isFeatureEnabled(const String& featureName) const;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -32,31 +32,162 @@
+static NSString * const tabIdKey = @"tabId";
+static NSString * const windowIdKey = @"windowId";
+
+
```
```suggestion

```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -32,31 +32,162 @@
+        parameters.enabled = [enabled boolValue];
+    else
+        // according to the documentation, this parameter defaults to "true"
+        parameters.enabled = true;
```
Defaults can be in the `WebExtensionSidebarParameters` struct. Comment isn't needed.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -32,31 +32,162 @@
+    if (auto path = objectForKey<NSString>(serializedParameters, @"path"))
+        parameters.panelPath = path;
+    else
+        parameters.panelPath = ""_s;
```
This isn't needed, or should be `panelPath { emptyString() }` in the struct. String defaults to `nullString()` if not specified otherwise, which is similar to empty ids that works too.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -32,31 +32,162 @@
+    auto result = parseTabIdentifier(options);
+    if ((*outExceptionString = indicatesError(result)))
+        return;
+    const auto tabId = variantTypeToOptional<WebExtensionTabIdentifier>(result);
```
```suggestion
        return;

    const auto tabId = variantTypeToOptional<WebExtensionTabIdentifier>(result);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -32,31 +32,162 @@
+    auto tabResult = parseTabIdentifier(options);
+    if ((*outExceptionString = indicatesError(tabResult)))
+        return;
+    auto tabId = variantTypeToOptional<WebExtensionTabIdentifier>(tabResult);
```
```suggestion
        return;

    auto tabId = variantTypeToOptional<WebExtensionTabIdentifier>(tabResult);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -32,31 +32,162 @@
+    auto windowResult = parseWindowIdentifier(options);
+    if ((*outExceptionString = indicatesError(windowResult)))
+        return;
+    auto windowId = variantTypeToOptional<WebExtensionWindowIdentifier>(windowResult);
```
```suggestion
        return;

    auto windowId = variantTypeToOptional<WebExtensionWindowIdentifier>(windowResult);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -32,31 +32,162 @@
+    }
+
+    WebProcess::singleton()
+        .sendWithAsyncReply(Messages::WebExtensionContext::SidebarOpen(windowId, tabId), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<void, WebExtensionError>&& result) {
```
```suggestion
    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::SidebarOpen(windowId, tabId), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<void, WebExtensionError>&& result) {
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+
+    if (maybeTabId && ![maybeTabId isKindOfClass:NSNumber.class])
+        return toErrorString(nil, @"details", @"'tabId' must be a number");
+    if (maybeWindowId && ![maybeWindowId isKindOfClass:NSNumber.class])
```
```suggestion
        return toErrorString(nil, @"details", @"'tabId' must be a number");

    if (maybeWindowId && ![maybeWindowId isKindOfClass:NSNumber.class])
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+template<typename VariantType>
+static std::tuple<std::optional<WebExtensionWindowIdentifier>, std::optional<WebExtensionTabIdentifier>> getIdentifiers(VariantType& variant)
+{
+    static_assert(isVariantMember<WebExtensionWindowIdentifier, VariantType>::value);
+    static_assert(isVariantMember<WebExtensionTabIdentifier, VariantType>::value);
```
```suggestion
    ASSERT(isVariantMember<WebExtensionWindowIdentifier, VariantType>::value);
    ASSERT(isVariantMember<WebExtensionTabIdentifier, VariantType>::value);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h:
```diff
@@ -35,6 +35,33 @@ OBJC_CLASS NSString;
+template<typename VariantType>
+NSString *indicatesError(const VariantType& variant)
+{
+    static_assert(isVariantMember<SidebarError, VariantType>::value);
```
```suggestion
    ASSERT(isVariantMember<SidebarError, VariantType>::value);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+        *outExceptionString = toErrorString(nil, @"details", @"'windowId' must be a number");
+        return;
+    }
+    std::optional<WebExtensionWindowIdentifier> windowId = maybeWindowId ? toWebExtensionWindowIdentifier(((NSNumber *) maybeWindowId).doubleValue) : std::nullopt;
```
```suggestion
    }

    std::optional<WebExtensionWindowIdentifier> windowId = maybeWindowId ? toWebExtensionWindowIdentifier(((NSNumber *) maybeWindowId).doubleValue) : std::nullopt;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+    auto result = parseSidebarActionDetails(details);
+    if ((*outExceptionString = indicatesError(result)))
+        return;
+    const auto [windowId, tabId] = getIdentifiers(result);
```
```suggestion
        return;

    const auto [windowId, tabId] = getIdentifiers(result);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+    auto panelResult = parseDetailsStringFromKey(details, panelKey);
+    if ((*outExceptionString = indicatesError(panelResult)))
+        return;
+    const auto panelPath = variantTypeToOptional<String>(panelResult);
```
```suggestion
        return;

    const auto panelPath = variantTypeToOptional<String>(panelResult);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+    auto result = parseSidebarActionDetails(details);
+    if ((*outExceptionString = indicatesError(result)))
+        return;
+    const auto [windowId, tabId] = getIdentifiers(result);
```
```suggestion
        return;

    const auto [windowId, tabId] = getIdentifiers(result);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+    auto result = parseSidebarActionDetails(details);
+    if ((*outExceptionString = indicatesError(result)))
+        return;
+    const auto [windowId, tabId] = getIdentifiers(result);
```
```suggestion
        return;

    const auto [windowId, tabId] = getIdentifiers(result);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+    auto titleResult = parseDetailsStringFromKey(details, titleKey);
+    if ((*outExceptionString = indicatesError(titleResult)))
+        return;
+    const auto title = variantTypeToOptional<String>(titleResult);
```
```suggestion
        return;

    const auto title = variantTypeToOptional<String>(titleResult);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,220 @@
+    auto result = parseSidebarActionDetails(details);
+    if ((*outExceptionString = indicatesError(result)))
+        return;
+    const auto [windowId, tabId] = getIdentifiers(result);
```
```suggestion
        return;

    const auto [windowId, tabId] = getIdentifiers(result);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h:
```diff
@@ -35,6 +35,33 @@ OBJC_CLASS NSString;
 
 namespace WebKit {
 
+using SidebarError = NSString *;
```
This needs to be a `String` or `RetainPtr<NSString>`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h:
```diff
@@ -35,6 +35,33 @@ OBJC_CLASS NSString;
+struct isVariantMember<T, std::variant<ALL_T...>> : public std::disjunction<std::is_same<T, ALL_T>...> { };
+
+template<typename OptType, typename... Types>
+std::optional<OptType> variantTypeToOptional(std::variant<Types...>& variant)
```
No need to name the argument type in the name, since it is part of the signature.
```suggestion
std::optional<OptType> toOptional(std::variant<Types...>& variant)
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h:
```diff
@@ -35,6 +35,33 @@ OBJC_CLASS NSString;
+// In this variant, `monostate` indicates that we have neither a window or tab identifier, but no error
+using ParseResult = std::variant<std::monostate, WebExtensionTabIdentifier, WebExtensionWindowIdentifier, SidebarError>;
+
+template<typename T, typename VARIANT_T>
+struct isVariantMember;
+template<typename T, typename... ALL_T>
+struct isVariantMember<T, std::variant<ALL_T...>> : public std::disjunction<std::is_same<T, ALL_T>...> { };
```
These generic helpers should go somewhere else, but I'm not sure where…

AriYoung00 replied:
I had the same dilemma. To me, they feel too general for the SidebarAction header, but not quite general enough for CocoaHelpers.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h:
```diff
@@ -35,6 +35,33 @@ OBJC_CLASS NSString;
+}
+
+template<typename VariantType>
+NSString *indicatesError(const VariantType& variant)
```
Can this move to the `mm` file?

AriYoung00 replied:
I don't think it can, since it's templated and I'm using it in multiple .mm files

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,339 @@
+
+    auto& sidebar = maybeSidebar.value().get();
+
+    if (sidebar.opensSidebar())
```
```suggestion
    auto& sidebar = maybeSidebar.value().get();
    if (sidebar.opensSidebar())
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,339 @@
+static Expected<T, WebExtensionError> toExpected(std::optional<T>&& opt, NSString * const errorMessage = @"value not found")
+{
+    if (opt)
+        return WTFMove(opt.value());
```
```suggestion
static Expected<T, WebExtensionError> toExpected(std::optional<T>&& optional, NSString * const errorMessage = @"value not found")
{
    if (optional)
        return WTFMove(optional.value());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,339 @@
+
+    auto& sidebar = maybeSidebar.value().get();
+
+    if (sidebar.canProgrammaticallyCloseSidebar())
```
```suggestion
    auto& sidebar = maybeSidebar.value().get();
    if (sidebar.canProgrammaticallyCloseSidebar())
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,339 @@
+    }
+    auto& sidebar = maybeSidebar.value();
+
+    if (!sidebar->canProgrammaticallyCloseSidebar()) {
```
```suggestion
    }

    auto& sidebar = maybeSidebar.value();
    if (!sidebar->canProgrammaticallyCloseSidebar()) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,339 @@
+    }
+    auto& sidebar = maybeSidebar.value();
+
+    if (sidebar->isOpen()) {
```
```suggestion
    }

    auto& sidebar = maybeSidebar.value();
    if (sidebar->isOpen()) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -32,55 +32,339 @@
 {
+    if (auto *controller = extensionController())
+        return controller->isFeatureEnabled(@"WebExtensionSidebarEnabled");
+
```
```suggestion
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -507,6 +511,20 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
+    }
+
+    return false;
+
```
```suggestion
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -32,52 +32,228 @@
+        return SidebarError { toErrorString(nil, @"details", [NSString stringWithFormat:@"'%@' must be of type 'string' or 'null'", key]) };
+    }
+
+    return SidebarError { (NSString *) maybeValue };
```
```suggestion
    return SidebarError { (NSString *)maybeValue };
```

AriYoung00 committed:
Fill in sidebarAction / sidePanel WebProcess <=> UIProcess messaging stubs.
https://webkit.org/b/277688
rdar://133073249

Reviewed by Timothy Hatcher.

This PR fills in most of the message senders and receivers which pertain
to sidePanel / sidebarAction. Additionally, it adds several helper
utilities to aid in parameter parsing for sidePanel / sidebarAction.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::toExpected): Utility function to convert an optional<T>
to an Expected<T, WebExtensionError>
(WebKit::getSidebarWithIdentifiers): Utility function to validate window
/ tab identifiers then return the sidebar object associated with said
identifiers
(WebKit::getOrCreateSidebarWithIdentifiers): Utility function to
validate window / tab identifiers and create a sidebar object
associated with said identifiers if necessary
(WebKit::WebExtensionContext::openSidebarForTab): Utility function to
open a sidebar object
(WebKit::WebExtensionContext::closeSidebarForTab): Utility function to
close a sidebar object
(WebKit::WebExtensionContext::sidebarOpen): Fill in message receiver for
sidebarOpen
(WebKit::WebExtensionContext::sidebarClose): Fill in message receiver for
sidebarClose
(WebKit::WebExtensionContext::sidebarIsOpen): Fill in message receiver for
sidebarIsOpen
(WebKit::WebExtensionContext::sidebarToggle): Fill in message receiver for
sidebarToggle
(WebKit::WebExtensionContext::sidebarSetIcon): Update stub with FIXME
(WebKit::WebExtensionContext::sidebarGetTitle): Fill in message receiver for
sidebarGetTitle
(WebKit::WebExtensionContext::sidebarSetTitle): Fill in message receiver for
sidebarSetTitle
(WebKit::WebExtensionContext::sidebarGetOptions): Fill in message receiver for
sidebarGetOptions
(WebKit::WebExtensionContext::sidebarSetOptions): Fill in message receiver for
sidebarSetOptions
(WebKit::WebExtensionContext::isSidebarMessageAllowed): Gate sidebar
messages behind sidebar feature flag
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
(WebKit::parseTabIdentifier): Utility function to parse and validate the
tab identifier received through a sidePanel details object
(WebKit::parseWindowIdentifier): Utility function to parse and validate
the window identifier received through a sidePanel details object
(WebKit::serializeSidebarParameters): Utility function to serialize a
WebExtensionSidebarParameters struct to an NSDictionary
(WebKit::deserializeSidebarParameters): Utility function to deserialize
a WebExtensionSidebarParameters struct from an NSDictionary
(WebKit::WebExtensionAPISidePanel::getOptions): Fill in message sender
for getOptions
(WebKit::WebExtensionAPISidePanel::setOptions): Fill in message sender
for setOptions
(WebKit::WebExtensionAPISidePanel::getPanelBehavior): Update stub with
FIXME
(WebKit::WebExtensionAPISidePanel::setPanelBehavior): Update stub with
FIXME
(WebKit::WebExtensionAPISidePanel::open): Fill in message sender for
open
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
(WebKit::parseSidebarActionDetails): Utility function to parse the
`details` object passed to many sidebarAction method calls
(WebKit::parseDetailsStringFromKey): Utility function to validate the
string value of some key in a dictionary, optionally allowing it to be
NSNull
(WebKit::getIdentifiers): Utility function to retrieve whatever window
or tab identifiers are set in a variant, returning them as
std::optionals
(WebKit::WebExtensionAPISidebarAction::open): Fill in message sender for
open
(WebKit::WebExtensionAPISidebarAction::close): Fill in message sender
for close
(WebKit::WebExtensionAPISidebarAction::toggle): Fill in message sender
for toggle
(WebKit::WebExtensionAPISidebarAction::isOpen): Fill in message sender
for isOpen
(WebKit::WebExtensionAPISidebarAction::getPanel): Fill in message sender
for getPanel
(WebKit::WebExtensionAPISidebarAction::setPanel): Fill in message sender
for setPanel
(WebKit::WebExtensionAPISidebarAction::getTitle): Fill in message sender
for getTitle
(WebKit::WebExtensionAPISidebarAction::setTitle): Fill in message sender
for setTitle
(WebKit::WebExtensionAPISidebarAction::setIcon): Fill in message sender
for setIcon
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h:
(WebKit::toOptional): Utility function to attempt to access
one particular field from a variant and return it in an std::optional,
returning std::nullopt if the variant does not contain the requested
field
(WebKit::indicatesError): Utility function to determine whether a
variant indicates that an error occurred, and return the error if it
does
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::isFeatureEnabled const): Helper
function to check if a particular feature is enabled using its name.

Canonical link: https://commits.webkit.org/282026@main
